### ISSUE TITLE
Developer site

### ISSUE THREAD

**Issue by ghost** [turtlecoin/meta#58]
Continuing from a separate discussion [here](https://github.com/turtlecoin/meta/issues/47), I've attempted to create a simple site (plain Jekyll) which houses the Guides and API docs.

**Site**: https://ac1984.github.io/turtle-docs/guides/
**Source**: https://github.com/ac1984/turtle-docs/

I would like some feedback on this and if it would be useful to the community at **this point** in time. It currently uses dummy content as my focus has been on setting up the site and structure. Based on the brand guide.

**Comment by SoreGums** [comment]
Looks great - I guess the advantage over the github wiki is flexibility in content and able to theme as per the TurtleCoin brand ;)

Hopefully people like it and adopt it or something very close to it. 

It is a very accessible for people to contribute that aren't necessarily developers. Create a GitHub account, fork, edit in the browser, PR :)

**Comment by RocksteadyTC** [comment]
I like this a lot. @BebopTC

**Comment by finnious** [comment]
@ac1984 That site looks great!
I already see a section I need to walk through :)

**Comment by ac1984** [comment]
I think the idea with this is to create a developer site – which includes Guides, API docs etc, which is a part of the turtlecoin website (like https://guides.github.com). Not just a second frontend to the GitHub wiki repo. If the idea is to actually use the GitHub wiki then lets just use that itself. @BebopTC Let me know what you think :)

**Comment by ac1984** [comment]
@BebopTC @RocksteadyTC  If we decide to stay with the wiki thats perfect too. This meta article solely is for something which is a little broader in scope. Its perfectly fine if we don't want to do it now and instead focus on the Github Wiki. Then I would focus my efforts on the wiki instead.

However, if we do decide to go with this then I see the work going like this:

1. Merge `docs`, `turtle-wiki` into one repo. 
2. Work with a volunteer to import the Bytecoin API docs in that repo.
3. Work with the core members to collect the material needed based on [_Sidebar.md](https://github.com/turtlecoin/turtlecoin-wiki/blob/master/_Sidebar.md).
4. Work with a volunteer copy writer to get to a consistent language and article structure.
5. Work with a designer to tweak the API and guide design - consistent gutters, typography adjustments, etc. for the site.

Later:

* Developer landing page: A page to tie everything together, like https://www.stellar.org/developers
* Blog: Maybe we just stay on Medium, that OK. Though, this would lets us collaborate on articles via Github. (like [Forknote's](http://forknote.net/blog/)).
* Have an equivalent of https://turtleturtle.org. This would be managed via Github like API docs and guides and would go through the same review process.
* More tooling - spelling check, broken link check.


Let me know what you think.

**Comment by RocksteadyTC** [comment]
@BebopTC I support this. What do you think? We need to get WIKI moving like yesterday.

**Comment by BebopTC** [comment]
@ac1984 we're about on the same page. I've already began working on pulling he bytecoin docs for RPC and `docs` was merged into `turtlecoin-wiki` before but may not be current.

The main desire here w/ turtlecoin-wiki, to be very clear, is not to leverage the github wiki feature (though we get that for free). Its purpose is to be **THE** central repository of **all** official documentation. It is the **single source of truth** and any site that has technical documentation needs to pull its content directly from turtlecoin-wiki moving forward, preferably in some kind of dynamic/on-the-fly way or if done statically, on a very frequent basis. This ensures consistency throughout our ecosystem and also makes editing/updating documentation simple, as there will only be one place to do it. 

Downstream presentations of the content can take advantage of the organized structure we'll follow and the consistent format (github style markdown). This is a familiar format to many developers and easy for newcomers to pick up, and can be edited directly in github even, which facilitates more contribution from more people in an easy way that fits in with our preferred commit-based workflow.

I'd be happy to rename it from turtlecoin-wiki to turtlecoin-docs or something if that mkes the meaning more apparent. I gave it the name turtlecoin-wiki because I initially cloned it from `turtlecon/turtlecoin.wiki.git` (the repo of the wiki of the core turtlecoin repo). Also, the main inspiration is the [ArchLinux Wiki (https://wiki.archlinux.org/)](https://wiki.archlinux.org), which I confidently believe is the best documentation of any software project in the history of the multiverse. Please take a look around the arch linux wiki if you haven't before, it is a beautiful thing to behold. The styling is plain and basic but sufficient; the content and structure is consistent, thorough, well organized and quite approachable, and there is little fluff.

**Comment by ar-x** [comment]
@BebopTC I see that the GH wiki is being overhauled via https://github.com/turtlecoin/turtlecoin-wiki/pull/100. Is a ArchiLinux/MediaWiki style system based on Jekyll still fair game? The idea of course is that the wiki would live in its own repo like today, and this system would pull in the markdown from there (via git submodules or something). The wiki articles might need a little more organizing to perhaps match the category/article nesting in this system. This system could also include some one off pages. 

Having said that, perhaps the most simple (but perhaps not the best technically) approach is to keep the Jekyll site and the wiki content in one repo. Like https://github.com/circleci/circleci-docs/. Would that be acceptable as a first cut? Thoughts?


cc @Sajo811

**Comment by Sajo811** [comment]
Yep, I'm onboard with this idea. I enquired about it a few weeks ago, am currently playing around with it a bit. The current wiki is too bare, using GitHub docs would let us use a custom domain and use custom branding. 

I also agree with having the Jekyll site and content in one repo, makes it easier to manage.

**Comment by ar-x** [comment]
@Sajo811 Great!

I'll just wait on a confirmation from @RocksteadyTC @BebopTC before spending time on the implementation.

**Comment by SoreGums** [comment]
From @BebopTC's comments above. the point is the repo is the central source of truth.
It gets access to Jekyll presentation for free by being a GitHub hosted git repo.

Basically, this means whatever presentation software that can consume GitHub flavoured markdown is fair game. Being markdown content, it can also be directly edited on the GitHub site.

From what I can tell the wiki and repo are the same, it simply takes a manual (at the moment) operation to publish the repo to the wiki. I'm [writing up a guide now ](https://github.com/SoreGums/turtlecoin-wiki/wiki/How-to-contribute-to-the-TurtleCoin-wiki) to show people how this all works. I only figured it out while writing the RainBorg guide.

CircleCi probably wanted to make the docs a bit prettier than what GitHub offers so set up a standalone git > generate > host workflow. Anyone can do that if they want, and is the point of using a central single point of truth, others can make things from it however they want. If they get bored of doing that, it is fine because the source is here and was not reliant on the other thing that person made.

**Comment by ar-x** [comment]
@SoreGums Just so that I understand you correctly.

> From @BebopTC's comments above. the point is the repo is the central source of truth.
It gets access to Jekyll presentation for free by being a GitHub hosted git repo.

The Jekyll presentation is only auto built by Github pages only if the Jekyll site and its markdown content are in the same repo. Is that OK? Also, do you think we should maintain the Github wiki as well as an official frontend to the markdown content? If the Github wiki is not to be used, then users would only need to create a simple PR and merge it into the `master` - all via the Github UI. The difference being that they would not have to deal with the wiki repo (or git CLI), and the publish 'site' step would be done automatically by Github.

Let me know what you think.

**Comment by ar-x** [comment]
If the Jekyll site and the markdown content are in the same repo, then the Github Wiki structure convention will not match the Jekyll conventions. So we'll be able to support either - unless we do something very elaborate.

For example, the Github Wiki needs to be like how it is today, whereas Jekyll would need to be something like this:

```
_about:
  Background-and-History.md
  Community.md
  Contributing.md
  Core-Developers.md
  Supporting.md
  Technical-Data.md

./_faqs:
  Frequently-Asked-Questions.md

./_getting-started:
  Getting-Started.md
  Setting-up-a-new-wallet.md
  setting-up-a-new-miner.md
```

So my main question here is – Does the structure need to be backward compatible with the Github wiki?, i.e. it would need to work on both – as a Github Wiki as well as the source for the Jekyll site.

My personal opinion is that the Github wiki organization doesn't scale so well with a large number of articles. And in the presence of a Jekyll site, its probably best not used.

Let me know what you all think.

**Comment by SoreGums** [comment]
there are two things.
GitHub code tab
GitHub wiki tab

Both of those tabs are separate git repositories.  
right now a manual step needs to be done to push master from code repo to master on wiki repo.

regarding Jekyll vs GitHub pages Jekyll layout. I don't understand any issues on the difference from an implementation point of view

This is what I see.  
GitHub Pages wiki = doesn't care about folders, can get to pages as long as they have unique file names. We can structure folders however we want. Unless I'm wrong on this point, I think the organisation problem is a non-issue.

If filenames are unique, then it doesn't matter if someone wants to run a pretty version with standalone Jekyll. single source of truth doesn't deviate, so can apply transformations to make it fit whatever. transforms would be pattern recognition and thus automatic. need to add .md to links, then do it, need to put folders in URL links then do it.

setting up a Jekyll site is not simple, it is like me saying just use git to edit the docs, why the problem? so if someone is going to put in the effort of putting up a pretty jekyll site, they can also put in the effort of applying a transformation and publishing that for anyone else to use as well - it probably already exists.  

no one can directly edit the official turtlecoin-wiki wiki tab repo because that would cause all kinds of problems. if people want to submit changes they need to edit the GitHub code repo, which they can do in the browser for simple edits (GitHub site provides a preview, PR gets created, everything is fine). If they need to go deeper then I'm writing a [guide ](https://github.com/SoreGums/turtlecoin-wiki/wiki/How-to-contribute-to-the-TurtleCoin-wiki)to help with that. It is a barrier to entry, however, the guide is supposed to smooth that barrier if the person doesn't want to do that well, #dev_docs exists, I'm sure they could team up with someone else and work together...

using git is not terribly difficult, granted it is not as easy as opening word or google docs and tapping on the keys. However, we gain a lot by being based on git and that lot outweighs people's non-desires to follow a guide.

that was pretty long - I hope I covered everything if not, let's discuss!
points
* 2 repos - not 1
* folders/directories don't matter to GitHub wiki generator, so just use unique file names whatever structure make sense organising wise
* can transform content to fit anything, the structure is known

**Comment by ar-x** [comment]
@SoreGums you are the best!, just need some more clarity to ensure I understand what you are saying.

> Both of those tabs are separate git repositories.

@SoreGums Are you saying that we keep the contents of the main repo frontend agnostic, and then use transforms to translate that content to a frontend specific format - Github Wiki and Jekyll (on a different repo too)?

> GitHub Pages wiki = doesn't care about folders, can get to pages as long as they have unique file names. We can structure folders however we want. Unless I'm wrong on this point, I think the organisation problem is a non-issue.

Ok got it. My current instinct is to turn the main repo to use a Jekyll like structure (for example for collection names), and then, if needed, update the transforms which for the wiki repo.

> setting up a Jekyll site is not simple, it is like me saying just use git to edit the docs, why the problem? so if someone is going to put in the effort of putting up a pretty jekyll site, they can also put in the effort of applying a transformation and publishing that for anyone else to use as well - it probably already exists.

I completely missed this one. What transformation are you talking about, and what should already exist? 

So I was initially thinking about something like this:

1. Update the repo structure to what Jekyll needs (preserving the git history ofcourse).
2. Build a pretty pretty Jekyll site based on that. 
3. Have Github pages auto build the site every time a PR merges into `master`. 

The repo/app on the first comment did all this, so its going to be very similar to that. @BebopTC had specific MediaWiki/ArchLinux thoughts on the design - I think we can to that over time.

Let me know what you think.

**Comment by SoreGums** [comment]
been chatting with @ar-x in #dev_general - cleared everything up.

ended with @BebopTC's quote above:
    * GitHub code tab - the repo is what matters, should be organised in a way that works for the content
    * GitHub wiki is rendering the files now for free, not the target
    * Any downstream presentation's of the repo are welcome, hopefully, they automatically update as changes are applied to repo

**Comment by ar-x** [comment]
Yep, I'll be giving it a shot. What I'm thinking is:

1. Create a Jekyll project which has a git submodule to `turtlecoin-wiki/master`. The site itself would be turtle branded and all.
2. This project would also have a GH webhook listener to `turtlecoin-wiki/master`. It would simply trigger a Travis CI run on the Jekyll project when a GH webhook from `turtlecoin-wiki/master` is received.
3. The Jekyll project would automatically get the latest `turtlecoin-wiki/master` via the submodule in the CI run.
4. A series of transforms would run on the wiki content, to turn it into a structure understood by Jekyll. This would be fairly minimal if the docs are correctly organized in their own regard.
5. Once the transforms are done, CI would run `jekyll build` and then publish the build artifacts to Github Pages via the pages API.

**Comment by ar-x** [comment]
.. or something like that. Lets see how it turns when I start on it. Thanks!

**Comment by ar-x** [comment]
@SoreGums Is it acceptable to add YAML frontmatter to markdown? For example like [this](https://github.com/travis-ci/docs-travis-ci-com/blob/c2a5d3f824baffa3df8091686c6013756e2903bc/user/language-specific.md). My main use case here is to be able to build the page title (the head one), and whereever the page gets referenced, like in the sidebar. For example https://github.com/turtlecoin/turtlecoin-wiki/blob/master/guides/Using-trtlbot-plus-plus.md - which is to be called _Using trtlbot ++_.

**Comment by ar-x** [comment]
This goes completely against what has been said before, maybe just using a stock mediawiki via Github OAuth could be an option. Like [ArchiLinux](https://wiki.archlinux.org), or [ByteCoin](https://en.bitcoinwiki.org/wiki/Bytecoin).

**Comment by brandonlehmann** [comment]
Is this issue resolved with all the other docs sites? I think it is.

**Comment by Sajo8** [comment]
#98

**Comment by SoreGums** [comment]
I think #98 covers this.

A unified front end, if it  uses two different rendering engines and generates the frame around them then that's what it does.
